In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../../data/processed/diabetes_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 229474
Columns : 22


,Diabetes_binary,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,0.0,1.0,1.0,1.0,40.0,1.0,0.0,0.0,0.0,0.0,...,1.0,0.0,5.0,18.0,15.0,1.0,0.0,9.0,4.0,3.0
1,0.0,0.0,0.0,0.0,25.0,1.0,0.0,0.0,1.0,0.0,...,0.0,1.0,3.0,0.0,0.0,0.0,0.0,7.0,6.0,1.0
2,0.0,1.0,1.0,1.0,28.0,0.0,0.0,0.0,0.0,1.0,...,1.0,1.0,5.0,30.0,30.0,1.0,0.0,9.0,4.0,8.0
3,0.0,1.0,0.0,1.0,27.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,0.0,0.0,0.0,0.0,11.0,3.0,6.0
4,0.0,1.0,1.0,1.0,24.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,3.0,0.0,0.0,0.0,11.0,5.0,4.0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "Diabetes_binary"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 229474
Columns : 21

TARGET VARIABLE (y)
Rows : 229474


0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Diabetes_binary, dtype: float64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


Diabetes_binary
0.0    194377
1.0     35097
Name: count, dtype: int64


Class Percentage


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
HighBP,float64
HighChol,float64
CholCheck,float64
BMI,float64
Smoker,float64
Stroke,float64
HeartDiseaseorAttack,float64
PhysActivity,float64
Fruits,float64
Veggies,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (183579, 21)
y_train shape : (183579,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (45895, 21)
y_test shape  : (45895,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


Diabetes_binary
0.0    84.71
1.0    15.29
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (183579, 21)
X_test_scaled shape  : (45895, 21)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,Veggies,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,1.096408,-0.889095,0.205176,-0.395435,1.072035,-0.215734,-0.338755,0.601961,-1.256327,0.509271,...,-4.187359,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,1.13290,-0.350440,0.020709,0.529942
1,1.096408,1.124739,0.205176,0.786628,-0.932805,-0.215734,-0.338755,0.601961,-1.256327,0.509271,...,0.238814,-0.3204,0.375330,-0.455262,-0.517047,-0.477229,-0.88269,0.620283,-0.986375,-1.381726
2,-0.912069,1.124739,0.205176,-0.690951,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,-0.88269,0.620283,1.027793,1.007859
3,-0.912069,1.124739,0.205176,-0.986467,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,-0.563578,-0.455262,-0.517047,-0.477229,-0.88269,1.591006,0.020709,-0.425892
4,1.096408,-0.889095,0.205176,0.195596,-0.932805,-0.215734,-0.338755,0.601961,0.795971,0.509271,...,0.238814,-0.3204,1.314239,-0.455262,0.698722,2.095428,-0.88269,-0.350440,-0.986375,-0.425892


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


HighBP                  0.0
HighChol                0.0
CholCheck              -0.0
BMI                    -0.0
Smoker                 -0.0
Stroke                  0.0
HeartDiseaseorAttack    0.0
PhysActivity           -0.0
Fruits                 -0.0
Veggies                -0.0
HvyAlcoholConsump      -0.0
AnyHealthcare           0.0
NoDocbcCost             0.0
GenHlth                 0.0
MentHlth                0.0
PhysHlth                0.0
DiffWalk               -0.0
Sex                    -0.0
Age                     0.0
Education              -0.0
Income                 -0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


HighBP                  1.0
HighChol                1.0
CholCheck               1.0
BMI                     1.0
Smoker                  1.0
Stroke                  1.0
HeartDiseaseorAttack    1.0
PhysActivity            1.0
Fruits                  1.0
Veggies                 1.0
HvyAlcoholConsump       1.0
AnyHealthcare           1.0
NoDocbcCost             1.0
GenHlth                 1.0
MentHlth                1.0
PhysHlth                1.0
DiffWalk                1.0
Sex                     1.0
Age                     1.0
Education               1.0
Income                  1.0
dtype: float64

# MRMR

## 1. Baseline

In [11]:
!pip install mrmr-selection

In [14]:
# =============================================================================
# MRMR BASELINE FEATURE SELECTION
# TOP-K = 5, 10, 15, 20
# =============================================================================

from mrmr import mrmr_classif
import pandas as pd


# -------------------------------------------------------------------------
# Align indices
# -------------------------------------------------------------------------

X_train_mrmr = X_train_scaled.copy()
y_train_mrmr = y_train.copy()

X_train_mrmr = X_train_mrmr.reset_index(drop=True)
y_train_mrmr = y_train_mrmr.reset_index(drop=True)


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Store MRMR results
# -------------------------------------------------------------------------

mrmr_results = {}


# =============================================================================
# RUN MRMR
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"TOP {k} MRMR FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # MRMR feature selection
    # ---------------------------------------------------------------------

    selected_features = mrmr_classif(
        X=X_train_mrmr,
        y=y_train_mrmr,
        K=k
    )


    # ---------------------------------------------------------------------
    # Select features
    # ---------------------------------------------------------------------

    X_train_selected = X_train_scaled[
        selected_features
    ]

    X_test_selected = X_test_scaled[
        selected_features
    ]


    # ---------------------------------------------------------------------
    # Store results
    # ---------------------------------------------------------------------

    mrmr_results[k] = {

        "features": selected_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected
    }


    # ---------------------------------------------------------------------
    # Display
    # ---------------------------------------------------------------------

    print("\nSelected Features:")

    print(selected_features)


print("\n" + "=" * 70)
print("MRMR FEATURE SELECTION COMPLETED")
print("=" * 70)


TOP 5 MRMR FEATURES


100%|██████████| 5/5 [00:01<00:00,  3.78it/s]



Selected Features:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk']

TOP 10 MRMR FEATURES


100%|██████████| 10/10 [00:02<00:00,  4.09it/s]



Selected Features:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Age', 'Income', 'PhysHlth', 'HvyAlcoholConsump']

TOP 15 MRMR FEATURES


100%|██████████| 15/15 [00:04<00:00,  3.52it/s]



Selected Features:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Age', 'Income', 'PhysHlth', 'HvyAlcoholConsump', 'CholCheck', 'Stroke', 'PhysActivity', 'Education', 'Veggies']

TOP 20 MRMR FEATURES


100%|██████████| 20/20 [00:12<00:00,  1.60it/s]


Selected Features:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Age', 'Income', 'PhysHlth', 'HvyAlcoholConsump', 'CholCheck', 'Stroke', 'PhysActivity', 'Education', 'Veggies', 'Smoker', 'MentHlth', 'Sex', 'AnyHealthcare', 'Fruits']

MRMR FEATURE SELECTION COMPLETED


In [15]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [16]:
# =============================================================================
# MRMR MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


mrmr_model_results = []


for k in k_values:

    Xtr = mrmr_results[k]["X_train"]

    Xts = mrmr_results[k]["X_test"]

    features = ", ".join(
        mrmr_results[k]["features"]
    )


    for model_name, model in models.items():

        # -------------------------------------------------------------
        # Train
        # -------------------------------------------------------------

        model.fit(
            Xtr,
            y_train
        )


        # -------------------------------------------------------------
        # Predictions
        # -------------------------------------------------------------

        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # -------------------------------------------------------------
        # Store metrics
        # -------------------------------------------------------------

        mrmr_model_results.append({

            "Method": "MRMR",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# =============================================================================
# CREATE RESULTS DATAFRAME
# =============================================================================

mrmr_results_df = pd.DataFrame(
    mrmr_model_results
)


display(mrmr_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,MRMR,5,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk",0.708552,0.308374,0.728736,0.433364,0.791913
1,MRMR,5,Random Forest,"GenHlth, HighBP, BMI, HighChol, DiffWalk",0.703366,0.305629,0.738709,0.432372,0.786967
2,MRMR,5,XGBoost,"GenHlth, HighBP, BMI, HighChol, DiffWalk",0.691077,0.299927,0.764496,0.430831,0.794595
3,MRMR,10,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.714631,0.317631,0.754096,0.446987,0.807581
4,MRMR,10,Random Forest,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.794161,0.325825,0.323550,0.324684,0.723269
5,MRMR,10,XGBoost,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.709228,0.315137,0.768201,0.446931,0.808270
6,MRMR,15,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.714806,0.318198,0.756803,0.448024,0.809239
7,MRMR,15,Random Forest,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.822290,0.362515,0.213563,0.268782,0.751129
8,MRMR,15,XGBoost,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.710818,0.316767,0.770053,0.448883,0.810332
9,MRMR,20,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.714304,0.318260,0.760080,0.448659,0.810583


In [17]:
mrmr_results_df.to_csv(
    "../../../results/diabetes/baseline/mrmr_baseline_results.csv",
    index=False
)

## 2. DODA

In [18]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [20]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\Anandha Lakshmi\Downloads\Anaconda\python.exe
DODA: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\__init__.py
Selector: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\selector.py


In [19]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [23]:
# =============================================================================
# MRMR ADAPTER FOR DODA
# =============================================================================

import numpy as np


class MRMRAdapter:

    def __init__(self, K):
        self.K = K
        self.importance = {}

    def fit(self, X, y):

        # Run MRMR
        selected_features = mrmr_classif(
            X=X,
            y=y,
            K=self.K
        )

        # -------------------------------------------------------------
        # Convert MRMR ranking into importance scores
        # First feature = highest score
        # -------------------------------------------------------------

        n_features = len(selected_features)

        scores = np.zeros(len(X.columns))

        feature_to_index = {
            feature: i
            for i, feature in enumerate(X.columns)
        }

        for rank, feature in enumerate(selected_features):

            index = feature_to_index[feature]

            scores[index] = (
                n_features - rank
            ) / n_features

        # -------------------------------------------------------------
        # Store importance
        # -------------------------------------------------------------

        self.importance = dict(
            zip(
                X.columns,
                scores
            )
        )

        return self

    def get_importance(self):

        return self.importance

    def get_name(self):

        return "MRMR"

In [40]:
# ============================================================
# FIX INDEX ALIGNMENT FOR MRMR
# ============================================================

# Convert X and y to pandas objects
X_mrmr = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
).reset_index(drop=True)

y_mrmr = pd.Series(
    y_train
).reset_index(drop=True)


print("X shape:", X_mrmr.shape)
print("y shape:", y_mrmr.shape)

print("X index:", X_mrmr.index[:5])
print("y index:", y_mrmr.index[:5])

X shape: (183579, 21)
y shape: (183579,)
X index: RangeIndex(start=0, stop=5, step=1)
y index: RangeIndex(start=0, stop=5, step=1)


In [41]:
from mrmr import mrmr_classif

mrmr_ranked_features = mrmr_classif(
    X=X_mrmr,
    y=y_mrmr,
    K=X_mrmr.shape[1]
)

print("Full MRMR ranking:")
print(mrmr_ranked_features)

print("Number of features:", len(mrmr_ranked_features))

100%|██████████| 21/21 [00:08<00:00,  2.33it/s]

Full MRMR ranking:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Age', 'Income', 'PhysHlth', 'HvyAlcoholConsump', 'CholCheck', 'Stroke', 'PhysActivity', 'Education', 'Veggies', 'Smoker', 'MentHlth', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']
Number of features: 21


In [42]:
# =============================================================================
# MRMR + DODA FEATURE SELECTION
# =============================================================================

import numpy as np
import pandas as pd


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "clinical_weights.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# MRMR Adapter
# -------------------------------------------------------------------------

class MRMRAdapter:

    def __init__(self, ranked_features):

        self.ranked_features = ranked_features
        self.importance = {}


    def fit(self, X, y):

        n_features = len(self.ranked_features)

        # Convert MRMR ranking into decreasing scores
        # Rank 1 = highest score
        # Rank n = lowest score

        for rank, feature in enumerate(
            self.ranked_features,
            start=1
        ):

            self.importance[feature] = (
                n_features - rank + 1
            ) / n_features

        return self


    def get_importance(self):

        return self.importance


    def get_name(self):

        return "MRMR"


# -------------------------------------------------------------------------
# Store results
# -------------------------------------------------------------------------

mrmr_doda_results = {}


# =============================================================================
# RUN MRMR + DODA
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"MRMR + DODA — TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # Create MRMR adapter
    # ---------------------------------------------------------------------

    mrmr_adapter = MRMRAdapter(
        mrmr_ranked_features
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(

        operators=[
            mrmr_adapter
        ],

        provider=provider,

        fusion=fusion,

        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on training data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform test data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = (
        selector.get_selected_features()
    )


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store results
    # ---------------------------------------------------------------------

    mrmr_doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores":
            selector.raw_math_scores_,

        "math_scores":
            selector.math_scores_,

        "clinical_weights":
            selector.clinical_weights_,

        "final_scores":
            selector.final_scores_
    }


print("\n" + "=" * 70)
print("MRMR + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


MRMR + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000002095908F4D0>
DODA Pipeline Started
Running: MRMR
Operator Scores:
{'MRMR_1': {'GenHlth': 1.0, 'HighBP': 0.9523809523809523, 'BMI': 0.9047619047619048, 'HighChol': 0.8571428571428571, 'DiffWalk': 0.8095238095238095, 'HeartDiseaseorAttack': 0.7619047619047619, 'Age': 0.7142857142857143, 'Income': 0.6666666666666666, 'PhysHlth': 0.6190476190476191, 'HvyAlcoholConsump': 0.5714285714285714, 'CholCheck': 0.5238095238095238, 'Stroke': 0.47619047619047616, 'PhysActivity': 0.42857142857142855, 'Education': 0.38095238095238093, 'Veggies': 0.3333333333333333, 'Smoker': 0.2857142857142857, 'MentHlth': 0.23809523809523808, 'Sex': 0.19047619047619047, 'AnyHealthcare': 0.14285714285714285, 'Fruits': 0.09523809523809523, 'NoDocbcCost': 0.047619047619047616}}
Resolving scores from: 1 operators

Raw Mathematical Scores
{'GenHlth': 1.0, 'HighBP': 0.9523809523809523, 'BMI':

In [43]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = mrmr_doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
0,GenHlth,1.000000,1.000000
1,HighBP,0.952381,0.952381
2,BMI,0.904762,0.904762
3,HighChol,0.857143,0.857143
4,DiffWalk,0.809524,0.809524
5,HeartDiseaseorAttack,0.761905,0.761905
6,Age,0.714286,0.714286
7,Income,0.666667,0.666667
8,PhysHlth,0.619048,0.619048
9,HvyAlcoholConsump,0.571429,0.571429


In [44]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
0,GenHlth,1.0
1,HighBP,1.0
2,BMI,1.0
11,Stroke,1.0
5,HeartDiseaseorAttack,1.0
4,DiffWalk,0.9
3,HighChol,0.9
6,Age,0.9
8,PhysHlth,0.8
15,Smoker,0.6


In [45]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
0,GenHlth,1.000000
1,HighBP,0.952381
2,BMI,0.904762
3,HighChol,0.771429
5,HeartDiseaseorAttack,0.761905
4,DiffWalk,0.728571
6,Age,0.642857
8,PhysHlth,0.495238
11,Stroke,0.476190
10,CholCheck,0.209524


In [46]:
selector.get_selected_features()

['GenHlth',
 'HighBP',
 'BMI',
 'HighChol',
 'HeartDiseaseorAttack',
 'DiffWalk',
 'Age',
 'PhysHlth',
 'Stroke',
 'CholCheck']

In [47]:
comparison = selector.compare_scores()

comparison


,Feature,Math Rank,Normalized Math Score,Clinical Weight,DODA Rank,Final Score,Rank Change
0,GenHlth,1,1.0000,1.0,1,1.0000,0
1,HighBP,2,0.9524,1.0,2,0.9524,0
2,BMI,3,0.9048,1.0,3,0.9048,0
3,HighChol,4,0.8571,0.9,4,0.7714,0
4,HeartDiseaseorAttack,6,0.7619,1.0,5,0.7619,1
5,DiffWalk,5,0.8095,0.9,6,0.7286,-1
6,Age,7,0.7143,0.9,7,0.6429,0
7,PhysHlth,9,0.6190,0.8,8,0.4952,1
8,Stroke,12,0.4762,1.0,9,0.4762,3
9,CholCheck,11,0.5238,0.4,10,0.2095,1


# MRMR + DODA Feature Selection — Interpretation

## 1. Overview

MRMR (Minimum Redundancy Maximum Relevance) was used as the mathematical feature-selection operator within the DODA framework.

The MRMR ranking was first generated from the training data. DODA then integrated the MRMR scores with domain-specific clinical weights using the Hadamard fusion mechanism:

\[
S_{DODA}(f) = S_{MRMR}(f) \times W_{clinical}(f)
\]

where:

- \(S_{MRMR}(f)\) = normalized MRMR score of feature \(f\)
- \(W_{clinical}(f)\) = clinical importance weight assigned to feature \(f\)
- \(S_{DODA}(f)\) = final DODA score

Top-K features were then selected for \(K = 5, 10, 15,\) and \(20\).

---

## 2. MRMR Ranking Before Clinical Adjustment

The MRMR operator produced the following ranking:

1. GenHlth
2. HighBP
3. BMI
4. HighChol
5. DiffWalk
6. HeartDiseaseorAttack
7. Age
8. Income
9. PhysHlth
10. HvyAlcoholConsump
11. CholCheck
12. Stroke
13. PhysActivity
14. Education
15. Veggies
16. Smoker
17. MentHlth
18. Sex
19. AnyHealthcare
20. Fruits
21. NoDocbcCost

This represents the purely mathematical ranking produced by MRMR without incorporating clinical knowledge.

---

## 3. Effect of Clinical Knowledge

DODA modifies the MRMR scores using predefined clinical weights.

Several clinically important features received high weights:

- GenHlth → 1.0
- HighBP → 1.0
- BMI → 1.0
- HeartDiseaseorAttack → 1.0
- Stroke → 1.0
- HighChol → 0.9
- DiffWalk → 0.9
- Age → 0.9
- PhysHlth → 0.8

In contrast, features such as Income, HvyAlcoholConsump, Fruits, and NoDocbcCost received lower clinical weights.

For example:

- Income: 0.6667 → 0.20 after clinical adjustment
- HvyAlcoholConsump: 0.5714 → 0.1143
- Stroke: 0.4762 → 0.4762

This demonstrates that DODA does not simply preserve the mathematical ranking. It modifies feature priority according to domain knowledge.

---

## 4. Important Ranking Changes

The most important effect is observed around the middle of the MRMR ranking.

### Stroke

Stroke had:

- MRMR rank: **12**
- MRMR normalized score: **0.4762**
- Clinical weight: **1.0**
- DODA score: **0.4762**

Because Stroke received the maximum clinical weight, its score was completely preserved.

Consequently, Stroke moved from **rank 12 to rank 9** in the DODA ranking.

This is an important result because Stroke is clinically relevant to cardiovascular/diabetes-related risk, but its purely mathematical MRMR score was not sufficiently high to place it in the Top-10.

### Income

Income had:

- MRMR rank: **8**
- MRMR score: **0.6667**
- Clinical weight: **0.3**
- DODA score: **0.20**

Therefore, Income was strongly down-weighted by the clinical knowledge layer.

### HvyAlcoholConsump

HvyAlcoholConsump had:

- MRMR rank: **10**
- MRMR score: **0.5714**
- Clinical weight: **0.2**
- DODA score: **0.1143**

It therefore received substantially lower priority after clinical adjustment.

---

## 5. Top-K Feature Selection

### Top-5

The selected features were:

```text
GenHlth
HighBP
BMI
HighChol
HeartDiseaseorAttack

In [50]:
# =============================================================================
# MRMR + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


mrmr_doda_model_results = []


# =============================================================================
# EVALUATE EACH TOP-K
# =============================================================================

for k in k_values:

    Xtr = mrmr_doda_results[k]["X_train"]

    Xts = mrmr_doda_results[k]["X_test"]

    features = ", ".join(
        mrmr_doda_results[k]["features"]
    )


    # -------------------------------------------------------------------------
    # Evaluate each ML model
    # -------------------------------------------------------------------------

    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        # Predictions
        y_pred = model.predict(
            Xts
        )


        # Probability predictions
        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # ---------------------------------------------------------------------
        # Store results
        # ---------------------------------------------------------------------

        mrmr_doda_model_results.append({

            "Method": "MRMR + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# =============================================================================
# CREATE RESULTS DATAFRAME
# =============================================================================

mrmr_doda_results_df = pd.DataFrame(
    mrmr_doda_model_results
)


display(mrmr_doda_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,MRMR + DODA,5,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.714261,0.313140,0.727597,0.437843,0.793300
1,MRMR + DODA,5,Random Forest,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.715786,0.313986,0.724462,0.438098,0.790212
2,MRMR + DODA,5,XGBoost,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.703955,0.308133,0.751389,0.437042,0.796183
3,MRMR + DODA,10,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.712975,0.316954,0.759082,0.447186,0.806608
4,MRMR + DODA,10,Random Forest,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.767905,0.315901,0.444080,0.369182,0.715571
5,MRMR + DODA,10,XGBoost,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.706286,0.312909,0.769768,0.444948,0.809073
6,MRMR + DODA,15,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.714304,0.317476,0.754951,0.446984,0.809135
7,MRMR + DODA,15,Random Forest,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.821440,0.362873,0.221684,0.275228,0.748295
8,MRMR + DODA,15,XGBoost,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.709445,0.315235,0.767631,0.446933,0.810423
9,MRMR + DODA,20,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, HeartDiseaseor...",0.714304,0.318260,0.760080,0.448659,0.810583


In [51]:
mrmr_doda_results_df.to_csv(
    "../../../results/diabetes/doda/mrmr_doda_results.csv",
    index=False
)

In [52]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        mrmr_results_df,
        mrmr_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,MRMR,5,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk",0.708552,0.308374,0.728736,0.433364,0.791913
1,MRMR,5,Random Forest,"GenHlth, HighBP, BMI, HighChol, DiffWalk",0.703366,0.305629,0.738709,0.432372,0.786967
2,MRMR,5,XGBoost,"GenHlth, HighBP, BMI, HighChol, DiffWalk",0.691077,0.299927,0.764496,0.430831,0.794595
3,MRMR,10,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.714631,0.317631,0.754096,0.446987,0.807581
4,MRMR,10,Random Forest,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.794161,0.325825,0.323550,0.324684,0.723269
5,MRMR,10,XGBoost,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.709228,0.315137,0.768201,0.446931,0.808270
6,MRMR,15,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.714806,0.318198,0.756803,0.448024,0.809239
7,MRMR,15,Random Forest,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.822290,0.362515,0.213563,0.268782,0.751129
8,MRMR,15,XGBoost,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.710818,0.316767,0.770053,0.448883,0.810332
9,MRMR,20,Logistic Regression,"GenHlth, HighBP, BMI, HighChol, DiffWalk, Hear...",0.714304,0.318260,0.760080,0.448659,0.810583


In [54]:
for k in k_values:

    selector = mrmr_doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Age', 'Income', 'PhysHlth', 'HvyAlcoholConsump', 'CholCheck', 'Stroke', 'PhysActivity', 'Education', 'Veggies', 'Smoker', 'MentHlth', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']

DODA Ranking:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'HeartDiseaseorAttack', 'DiffWalk', 'Age', 'PhysHlth', 'Stroke', 'CholCheck', 'Income', 'PhysActivity', 'Smoker', 'MentHlth', 'HvyAlcoholConsump', 'Education', 'Veggies', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']

Top-K Same: False

TOP 10
Raw Math Ranking:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'DiffWalk', 'HeartDiseaseorAttack', 'Age', 'Income', 'PhysHlth', 'HvyAlcoholConsump', 'CholCheck', 'Stroke', 'PhysActivity', 'Education', 'Veggies', 'Smoker', 'MentHlth', 'Sex', 'AnyHealthcare', 'Fruits', 'NoDocbcCost']

DODA Ranking:
['GenHlth', 'HighBP', 'BMI', 'HighChol', 'HeartDiseaseorAttack', 'DiffWalk', 'Age', 'PhysHlth', 'Stroke', 'CholCheck'

# MRMR + DODA Results Interpretation

## 1. Overall Performance

The MRMR-based feature selection was evaluated using Logistic Regression, Random Forest, and XGBoost for Top-K feature subsets of 5, 10, 15, and 20 features.

The results show that **MRMR + DODA generally maintains or slightly improves the predictive performance of MRMR alone for Logistic Regression and XGBoost**, while Random Forest shows a different behavior.

---

## 2. MRMR Baseline

### Top 5 Features

The MRMR baseline selected:

- GenHlth
- HighBP
- BMI
- HighChol
- DiffWalk

The best performance among the three models was obtained by **Logistic Regression in terms of F1 score (0.4334)**, while XGBoost achieved the highest Recall (0.7645) and ROC-AUC (0.7946).

This indicates that the Top-5 MRMR features contain substantial predictive information, but the model still faces the challenge of class imbalance.

### Top 10–20 Features

Increasing the number of features generally improved the ROC-AUC of Logistic Regression and XGBoost.

For example, XGBoost achieved:

- Top 5: ROC-AUC = **0.7946**
- Top 10: ROC-AUC = **0.8083**
- Top 15: ROC-AUC = **0.8103**
- Top 20: ROC-AUC = **0.8118**

This suggests that additional features provide useful complementary information beyond the initial five MRMR-selected features.

Random Forest showed a different pattern. Its accuracy increased considerably with larger feature sets, but its Recall and F1 score decreased. Therefore, the high accuracy should not be interpreted as better predictive performance because the dataset is imbalanced.

---

## 3. MRMR + DODA

DODA incorporates clinical knowledge into the mathematical ranking produced by MRMR.

The MRMR ranking initially placed `Income` above `PhysHlth`, `HvyAlcoholConsump`, and `Stroke`.

After applying the clinical knowledge layer:

- `HeartDiseaseorAttack` moved upward.
- `Stroke` moved substantially upward.
- `PhysHlth` moved upward.
- `Income` was down-weighted because of its lower clinical relevance.
- Several lifestyle and socioeconomic variables received lower final scores.

For example, the MRMR ranking was:

`GenHlth → HighBP → BMI → HighChol → DiffWalk → HeartDiseaseorAttack → Age → Income → PhysHlth → HvyAlcoholConsump → CholCheck → Stroke`

After DODA:

`GenHlth → HighBP → BMI → HighChol → HeartDiseaseorAttack → DiffWalk → Age → PhysHlth → Stroke → CholCheck → Income → ...`

This demonstrates that DODA does not simply reproduce the mathematical ranking. Instead, it **reorders features according to both statistical evidence and clinical knowledge**.

---

## 4. Effect on Top-K Feature Selection

### Top 5

**MRMR:**

`GenHlth, HighBP, BMI, HighChol, DiffWalk`

**MRMR + DODA:**

`GenHlth, HighBP, BMI, HighChol, HeartDiseaseorAttack`

DODA replaces `DiffWalk` with `HeartDiseaseorAttack`.

This is an important result because both features have strong clinical relevance, but DODA prioritizes the cardiovascular history variable based on the assigned clinical knowledge.

---

### Top 10

**MRMR:**

`GenHlth, HighBP, BMI, HighChol, DiffWalk, HeartDiseaseorAttack, Age, Income, PhysHlth, HvyAlcoholConsump`

**MRMR + DODA:**

`GenHlth, HighBP, BMI, HighChol, HeartDiseaseorAttack, DiffWalk, Age, PhysHlth, Stroke, CholCheck`

The major changes are:

- `Income` moves down.
- `HvyAlcoholConsump` moves down.
- `Stroke` enters the Top 10.
- `CholCheck` enters the Top 10.

This demonstrates the intended behavior of the clinical adjustment layer.

---

## 5. Model-Level Comparison

### Logistic Regression

Logistic Regression produced relatively stable results across MRMR and MRMR + DODA.

For example:

| Top-K | MRMR ROC-AUC | MRMR + DODA ROC-AUC |
|------:|-------------:|--------------------:|
| 5 | 0.7919 | 0.7933 |
| 10 | 0.8076 | 0.8066 |
| 15 | 0.8092 | 0.8091 |
| 20 | 0.8106 | 0.8106 |

The differences are small, indicating that DODA **preserves the predictive information captured by MRMR** while changing the feature composition.

---

### XGBoost

XGBoost also shows very similar performance:

| Top-K | MRMR ROC-AUC | MRMR + DODA ROC-AUC |
|------:|-------------:|--------------------:|
| 5 | 0.7946 | 0.7962 |
| 10 | 0.8083 | 0.8091 |
| 15 | 0.8103 | 0.8104 |
| 20 | 0.8118 | 0.8118 |

The results are particularly encouraging because DODA changes the feature ranking while maintaining almost identical predictive performance.

This suggests that the clinically prioritized features can replace some statistically ranked features **without sacrificing predictive discrimination**.

---

### Random Forest

Random Forest behaves differently.

Its accuracy increases substantially at Top-15 and Top-20, but Recall and F1 decrease.

For example:

- MRMR Top-20: Accuracy = **0.8436**, Recall = **0.1571**, F1 = **0.2351**
- MRMR + DODA Top-20: Accuracy = **0.8423**, Recall = **0.1510**, F1 = **0.2265**

Therefore, the higher accuracy is mainly associated with predicting the majority class and should not be considered evidence of improved clinical performance.

For this imbalanced dataset, **Recall, F1 score, and ROC-AUC are more informative than accuracy alone**.

---

## 6. Key Finding

The most important finding from the MRMR experiment is that **DODA changes the feature ranking substantially while preserving predictive performance**.

The mathematical MRMR ranking favors features such as `Income` and `HvyAlcoholConsump` relatively highly. After incorporating clinical knowledge, DODA promotes clinically meaningful variables such as `HeartDiseaseorAttack`, `Stroke`, and `PhysHlth`.

Despite these changes, Logistic Regression and XGBoost maintain almost the same ROC-AUC values.

Therefore, the results provide preliminary evidence that:

> **DODA can introduce clinical knowledge into feature selection without substantially reducing predictive performance.**

This is an important result for the proposed clinically guided feature-selection framework.

---

## 7. Important Observation About Stroke

`Stroke` provides a particularly useful demonstration of the DODA mechanism.

Under MRMR, `Stroke` was ranked **12th**, which excluded it from the Top-10 feature set.

After applying DODA, its clinical weight of **1.0** increased its final score and moved it to **9th position**, allowing it to enter the Top-10 feature set.

This demonstrates that DODA can prevent clinically important variables from being excluded solely because their statistical ranking is relatively low.

---

## 8. Conclusion

Overall, the MRMR experiment supports the proposed DODA framework.

The results demonstrate three important behaviors:

1. **MRMR identifies statistically relevant and complementary features.**
2. **DODA modifies the ranking according to predefined clinical knowledge.**
3. **The clinical adjustment does not substantially reduce predictive performance.**

The similarity in ROC-AUC between MRMR and MRMR + DODA, particularly for Logistic Regression and XGBoost, suggests that the clinically guided feature sets retain most of the predictive information of the original MRMR selections.

However, these results should be considered **preliminary evidence**. Further validation using repeated cross-validation, feature-selection stability analysis, and comparison across multiple datasets is required before concluding that DODA consistently outperforms conventional feature-selection methods.